text/0000-cargo-rustup-discovery.md

```diff
@@ -0,0 +1,457 @@
+It might be nice for Cargo and Rustup to adopt something similar,
+but it is not a replacement for ownership checks since it is opt-in.
+
+## Honor the user of the current directory
+
+The ownership check could also include the owner of the current working directory as "approved" and allow any files owned by that user in any parent directory.
+This is under the assumption that the user running the command makes an explicit choice as to which directory they run Cargo or Rustup from.
+This could significantly ease the burden of dealing with this issue,
+as it will avoid needing to set safe directories in almost all use cases where they would be needed.
+
+However, I feel like this approach is too risky.
+For `git`, this would be risky because users often run `git` from the shell prompt.
+This means that just changing directory to another user's directory could expose them.
+This may also apply to Rust users.
+For example [Starship](https://starship.rs/) can execute `rustc` to get its version for the shell prompt.
+
+## Filesystem behavior risks
+
+This fix assumes that a user can't in any way create a file owned by another user
+(unless they have elevated permissions like root).
+I believe this is the case for all major operating systems and filesystems.
+
+One (major) known exception is if a network-mounted filesystem is configured to map all files to the local user, which is not uncommon.
+In this case, the user will need to trust all users who can write to the mounted filesystem.
+
+## Check the ownership of every directory while traversing upwards
+
+The Git developers considered checking the ownership of every directory while traversing upwards,
+but they decided not to do it for performance reasons.
+I don't see a particularly strong reason to check the ownership along the way.
+I'd also be concerned about the performance consequence, though I expect it is extremely small.
+
+## Safe directory issues with Windows, WSL, and mingw
+
+On Windows, Git users have experienced some confusion and problems with setting safe directories.
+This is because git-for-windows is a mingw application,
+and needs to straddle the difference between POSIX and Windows style paths,
+as well as network paths.
+
+There is some concern that Cargo users using WSL or git-for-windows bash will have difficulty or confusion on how to set paths for the safe directories variable.
+I do not think it should be too much of an issue for Cargo because Cargo never uses POSIX paths, and only uses Windows-style paths.
+If the error message makes the full path clear, then the user should be able to copy-paste it.
+
+This may be particularly difficult for a user running Windows `cargo.exe` from within WSL.
+Paths in this environment look like `\\wsl$\Ubuntu-20.04\home\eric\foo`.
+I don't know how common this use case is, as I would expect someone doing things in WSL would install the native rustup/cargo binaries instead of using the ones from Windows.
```

Comment:
That does sound like a tad pain. I even [wrote a build definition](https://github.com/nullpo-head/WSL-Hello-sudo/blob/master/Makefile#L3) that relies on being able to call the Windows installation of Cargo from WSL to make building easier.

Reply:
Interesting! I was not expecting anyone to be doing that, though I guess I'm biased by mostly using mingw instead.

I think the error message should provide something that should work to copy-and-paste, so I hope that should be sufficient. This is also something that we may be able to iterate on without breaking things.